### DISCUSSION TITLE
Any way to reuse mutations?

### DISCUSSION THREAD

**Discussion by Eduard-Hasa** [trpc/trpc#3220] [General]
I feel there is a lot of code when defining a mutation on trpc client end especially when defining things like invalidations. I can't help but think if I can create reusable mutations and just declare all of the mutations and even queries in a react hook and just use them as functions that are exported from the hooks? 

Is this totally wrong logic?

**Comment by sachinraja** [comment]
That sounds fine but could you show an example of what you're talking about?

**Comment by Eduard-Hasa** [comment]
<img width="1570" alt="image" src="https://user-images.githubusercontent.com/48266101/203112205-268a2b20-6c49-4301-8591-120725d0b16f.png">

@sachinraja  Within the hook file it seems to work fine but when the hook is imported in the page Its throwing errors and I seem to lose the types. What is the norm that everyone has been approaching? Do you guys just declare a mutation along with its invalidation options each time you use a mutation? Am I trying to hard to create reusable code here?

**Comment by nkalpak** [comment]
The way I interpret your question is that you want an abstraction over the TRPC React Query hook. So instead of using
```ts
trpc.menu.createMenu.useMutation({
  onSuccess: () => { // Invalidate some stuff here... }
})
```
across the app multiple times, you have a custom hook which implements the `useMutation` with the invalidation internally:
```ts
// usage file
useCreateMenu()

// definition file
function useCreateMenu() {
  const utils = trpc.useContext()
  return trpc.menu.createMenu.useMutation({
    onSuccess: () => { utils.menu.getMenus.invalidate() },
  })
}
```
This is because you always want to invalidate the `getMenus` query when a new menu is created, and you don't want to duplicate the invalidation code in every usage of `trpc.menu.createMenu.useMutation`.

I try to do the same thing in my applications, and I usually approach it like this:
```ts
// usage file
useCreateMenu({ staleTime: Infinity })

// definition file
function useCreateMenu(options?: CreateMenuOptions) {
  const utils = trpc.useContext()
  return trpc.menu.createMenu.useMutation({
    onSuccess: () => { utils.menu.getMenus.invalidate() },
    ...options
  })
}
```

I had some issues however with correctly inferring the `options` type for usage in these custom hooks, so I opened this [discussion](https://github.com/trpc/trpc/discussions/3287) and subsequent [PR](https://github.com/trpc/trpc/pull/3298) to address these. You might want to check those out as they also have some more details about this use case.

**Reply by Eduard-Hasa** [reply]
Wow thanks for the amazing response. Very helpful.
